content/posts/exactly-once-http.md

```diff
@@ -0,0 +1,177 @@
+# HTTP protocol for exactly-once processing
+
+This article proposes a HTTP-based protocol that can be used to ensure remote invocation is conducted exactly once. It is an adaptation of the [token-based-deduplication](https://exactly-once.github.io/posts/token-based-deduplication/) idea described previously.
+
+One of the most common system integration scenarios is executing a function remotely and fetching the result. It is easy if the target function is _pure_ (does not have any side effects). But what if it does?
+
+The basic building blocks of HTTP protocol and REST approach are not enough to guarantee exactly-once execution in such a scenario. Let's remind ourselves what operations we have on our disposal.
+
+### HTTP
+
+HTTP defines a set of operations (known as _verbs_) that help to convey the meaning of a given API call. They also have some specific constraints. The most commonly known and used are `GET`, `POST`, `PUT` and `DELETE`. 
+
+`GET` is expected to not cause any side effects in the target system. Of course the implementor of a `GET` call can be evil and update database but that would be a violation of the protocol. The reason why `GET` should not have side effects is the fact that the response to this operation type can be cached so the caller can't be sure (unless explicitly opts out from caching) if the response comes from the target server or a cache.
+
+Next there is `PUT` and `DELETE`. These operations are expected to have idempotent implementation. This means that the caller should be able to retry calling `PUT` or `DELETE` several times and the side effects should be exactly as if they called it once. Again, there is nothing that prevents an evil programmer from violating this protocol assumption. Moreover, there is nothing in the protocol or common HTTP libraries that helps the implementor to comply with the idempotence requirements.
+
+Last but not least, `POST` carries no idempotency guarantees whatsoever. This is why `POST` is frequently used when the implementation can't provide any such guarantees or the service provider could not care enough to provide them. 
+
+### Traditional solutions to HTTP deduplication
+
+There are two well-known methods of ensuring the operation requested via a HTTP call is performed exactly once. 
+
+#### GET-before-POST
+
+This approach places the entire burden of deduplication on the caller. The caller is supposed to query the state of the service via `GET` operations to ensure that a given operation it is about to invoke has not yet been invoked. The downside of this approach is that if might be difficult to obtain the necessary information via regular `GET` calls. It might be necessary to provide a dedicated `GET` API just for that purpose. Another issue with this approach is the fact that it requires serializing the calls to a given service in order to guarantee correctness. For these reasons most people choose a different method.
```

Comment:
```suggestion
This approach places the entire burden of deduplication on the caller. The caller is supposed to query the state of the service via `GET` operations to ensure that a given operation it is about to invoke has not yet been invoked. The downside of this approach is that it might be difficult to obtain the necessary information via regular `GET` calls. It might be necessary to provide a dedicated `GET` API just for that purpose. Another issue with this approach is the fact that it requires serializing the calls to a given service in order to guarantee correctness. For these reasons, most people choose a different method.
```